api/client/authn.go

```diff
@@ -0,0 +1,108 @@
+package client
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/pkg/term"
+)
+
+// ValidateAuthnOpt checks if a passed-in option value is a recognized
+// authentication option.
+func ValidateAuthnOpt(option string) (string, error) {
+	if strings.HasPrefix(option, "basic.username=") ||
+		strings.HasPrefix(option, "basic.password=") ||
+		strings.HasPrefix(option, "interactive=") ||
+		strings.HasPrefix(option, "bearer.token=") {
+		return option, nil
+	}
+	return "", fmt.Errorf("invalid authentication option %s", option)
+}
+
+func (cli *DockerCli) getBasic(challenge string, defusername string) (string, string, error) {
+	username, ok := cli.authnOpts["basic.username"]
+	if !ok {
+		username = os.Getenv("DOCKER_AUTHN_USERID")
+	}
+	if username == "" {
+		username = defusername
+	}
+	defusername = username
+	password, ok := cli.authnOpts["basic.password"]
+	if !ok {
+		password = os.Getenv("DOCKER_AUTHN_PASSWORD")
+	}
+	interactive, ok := cli.authnOpts["interactive"]
+	if !ok {
+		interactive = "true"
+	}
+
+	if username == "" || password == "" {
+		if !cli.isTerminalIn || !cli.isTerminalOut {
+			logrus.Debugf("not connected to a terminal, not prompting for Basic auth creds")
+			return "", "", nil
+		}
+		if prompt, err := strconv.ParseBool(interactive); !prompt || err != nil {
+			logrus.Debugf("interactive prompting disabled, not prompting for Basic auth creds")
+			return "", "", nil
+		}
+
+		readInput := func(in io.Reader, out io.Writer) string {
+			reader := bufio.NewReader(in)
+			line, _, err := reader.ReadLine()
+			if err != nil {
+				fmt.Fprintln(out, err.Error())
+				os.Exit(1)
+			}
+			return string(line)
+		}
+
+		if defusername != "" {
+			fmt.Fprintf(cli.out, "Username[%s]: ", defusername)
+		} else {
+			fmt.Fprintf(cli.out, "Username: ")
+		}
+		username = readInput(cli.in, cli.out)
+		username = strings.Trim(username, " ")
+		if username == "" {
+			username = defusername
```

Comment:
Yep.  The idea was to take a default from the client library, replacing it with a supplied value if one was supplied at the command line or via the environment.  If no password was similarly supplied, the client would only need to confirm the user name value.